Chjune android step1

engineer-choi opened the pull request:
첫 번째 과제 제출합니다.
테스트코드 통과는 하게 겁나 하드코딩했는데 객체지향 생활 체조 원칙 지키는게 여간 쉬운게 아니네요.
for 안의 when 안의 if 안의 when 보기만해도 숨막힙니다만 일단 결과는 냈으니 제출은 합니다.
리팩터링 겁나해야겠네요 ㅎ_ㅎ

engineer-choi committed:
[First Commit]
TDD 연습

engineer-choi committed:
[First Commit]
TDD 연습

engineer-choi committed:
[테스트입니다]
Step1 브랜치 파서 푸시 및 브랜치 생성 테스트입니다

engineer-choi committed:
[예외처리 테스트코드 완료]
사칙연산 기호가 아닌 &를 사용하였을 때 IllegalArgumentException 발생하도록 예외처리 완료

engineer-choi committed:
[사칙연산 테스트코드 완료]
+ - * / 사칙연산에 대한 테스트코드 작성 완료

engineer-choi committed:
[결과도출 완료]
+ - * / 사칙연산에 대한 결과값 출력 완료.
겁나 하드코딩이라 리팩터링 개많이 필요할듯

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,61 @@
+import java.lang.IllegalArgumentException
+
+fun main() {
+    val inputData = readLine() ?: throw IllegalArgumentException()
```
inputData가 어떤 데이터일까? "입력 값" 이란 말은 너무 추상적이라 생각해.

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,61 @@
+        throw IllegalArgumentException("사칙연산 기호 쓰셔야 합니다!!!")
+    }
+    calculate(splitData)
+}
```
한 메서드가 여러 일을 하고 있어. 책임을 분리해보자.
1. 연산기호가 아닌 문자를 검증하는 로직
2. 계산로직

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,61 @@
+                }
+            }
+        }
+    }
```
인덴트 3 초과!!! 메서드로 더 쪼개보자.

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,61 @@
+
+fun sumData(a: Int, b: Int): Int {
+    return a + b
+}
```
클린 코드 서적에 "불용어 (noise word)를 추가하는 방식은 적절하지 못하다" 라는 말이있어.
Data, Info, The 같은 단어들을 불용어 불러. 의미가 불분명해지고 의도를 알 수 없기 때문이라고 생각해. 또, Data 라는 말이 없어도 말이 되는 경우가 많기 때문이야!

malibinYun commented on src/test/kotlin/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,62 @@
+    fun checkCalculableTest() {
+        assertThatThrownBy { checkCalculable(listOf("&")) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("사칙연산 기호 쓰셔야 합니다!!!")
```
나는 hasMessage("") 만 써봤는데 이것도 되나보네! 굳굳👍👍

malibinYun commented on src/test/kotlin/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,62 @@
+            .hasMessageContaining("사칙연산 기호 쓰셔야 합니다!!!")
+    }
+
+    @Test
+    fun calculateTest() {
+        assertThat(calculate(listOf("1", "+", "7", "-", "2", "*", "3", "/", "2")))
+            .isEqualTo(9)
+    }
```
계산 테스트는 아마 최종 사용 테스트를 의미하는거같아. 
사용자는 split 된 list를 넘겨주지 않으니, 통짜 문자열을 가지고 테스트 해보는건 어떨까?

malibinYun commented on src/test/kotlin/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,62 @@
+        assertThat(1).isEqualTo(changeToInt("1"))
+    }
+
+    @Test
+    fun sumDataTest() {
+        assertThat(1 + 2).isEqualTo(sumData(1, 2))
+    }
```
```suggestion
    @Test
    fun sumDataTest() {
        assertThat(sumData(1, 2)).isEqualTo(3)
    }
```
이런식으로 순서를 바꾸는 게 좀 더 가독성이 좋아보여!
that 이하를 assert 하겠다 라는 뜻이니까 실제 대상보다 비교 대상이 먼저 오는게 맞아보여!

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,61 @@
+    return a + b
+}
+
+fun minusData(a: Int, b: Int): Int {
+    return a - b
+}
+
+fun multiplyData(a: Int, b: Int): Int {
+    return a * b
+}
+
+fun divideData(a: Int, b: Int): Int {
+    return a / b
```
a - b 를 minusData 로 래핑하는 것 보다
a - b 라고 쓰는게 더 가독성이 좋아보여. 메서드로 래핑한게 오버피팅이 된 느낌이야.

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,61 @@
+    return resultNum
+}
+
+fun changeToInt(number: String): Int {
+    return number.toInt()
+}
```
String.toInt() 확장함수만 쓰는게 이렇게 래핑하는 것 보다 더 가독성이 좋아보여. 오버피팅 느낌이랄까?

malibinYun reviewed (changes requested):
1등 과제 제출 감사합니다. 😊😊

현재 아래 사항들을 잡아내지 못하고 있어.
* "2 3" 입력 시 결과 값이 2
* "2 +" 입력 시 인덱스 초과 Exception 발생
*  "3 # 4" 입력 시 결과 값이 3
  * & 는 걸러내지만, 연산 기호가 있을 자리에 사칙 연산기호 이외의 것을 거르지 못함.

사실 이 프로그램이 제대로 동작하는게 먼저라서, 리뷰 내용보다는 이런 예외사항을 잡아주는게 더 중요할 거 같아. 이런 놈들을 먼저 잡아 낸 뒤, 리팩터링을 진행해도 좋을 거 같아.

전체적으로 모든 메서드가 전역으로 설정되어있네. 이렇게 하면 어떤 단점이 있을까? 우선은 모든 메서드를 전역으로 만들고 하는것도 좋은 방법중 하나야. 이런 방법을 Top-Down 방식이라 하는데, 스터디 때 좀 더 설명해볼게.

전역으로 선언되어있는 메서드들을 잘 보고, 책임이 비슷한 애들끼리 묶어서 클래스를 만들어보는건 어떨까? 그럼으로써 클래스의 책임과 역할을 부여해보는거야.

engineer-choi committed:
[1차 리팩터링 완료]
피드백 받은 예외처리 추가.
전역 메서드 -> 클래스별 역할 분배

engineer-choi reviewed (commented):
checkCalculable에서 private로 갖고 있는 두 가지 메서드를 테스트하도록 만들었습니다만.
테스트 코드에서 검사하는 결과는 CalculateChecker가 private로 갖고 있는 두 메소드의 Exception메시지를 받는지 처리하도록 했는데 섞인 느낌이 없잖아 있네요.. 피드백 부탁드립니다

malibinYun commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -0,0 +1,24 @@
+        checkOperatorNumber(operatorList, numberList)
+    }
+
+    private fun checkOperatorNumber(operatorList: List<String>, numberList: List<String>) {
+        require(operatorList.size + 1 == numberList.size) { IllegalArgumentException("연산자 : $operatorList 숫자 : $numberList") }
+    }
+
```
연산자 리스트와 숫자 리스트만 띡 보여주면 나중에 봤을 때 뭘 의도했는지 까먹지 않을까?

malibinYun commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -0,0 +1,24 @@
+import java.lang.IllegalArgumentException
+
+class CalculateChecker {
+    private val fourRulesCalculationList = listOf("+", "-", "*", "/")
```
상수처리! (static, 대문자 snake case 등)

malibinYun commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -0,0 +1,24 @@
+class CalculateChecker {
+    private val fourRulesCalculationList = listOf("+", "-", "*", "/")
+    fun checkCalculable(splitList: List<String>) {
+        val operatorList = splitList.filterNot { it.toIntOrNull() is Int }
+        checkOperatorList(operatorList)
+        val numberList = splitList.filter { it.toIntOrNull() is Int }
+        checkOperatorNumber(operatorList, numberList)
```
2 3 4 가 들어와도 operaorList에는 아무것도 쌓이지 않지만, 억지로 validation을 하는 느낌이 들어.
String을 nullable한 Int로 파싱을 유도 함으로써 나는 요구사항을 알기 때문에 뭘 하려는건진 알겠지만.. 해석하는데 시간이 들거 같다는 생각이랄까?
아예 인덱스 기반이나 순서 기반으로 검증 해도 될거같아. 꼭 파싱이 되는지 안되는지의 유무를 떠나서

malibinYun commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    private fun checkOperatorList(operatorList: List<String>) {
+        for (i in operatorList.indices) {
+            val operator = operatorList[i]
+            require(fourRulesCalculationList.contains(operator)) {
+                IllegalArgumentException("${operator}는 사칙 연산기호가 아닙니다.")
+            }
+        }
```
```suggestion
        for (operator in operatorList) {
            require(fourRulesCalculationList.contains(operator)) {
                IllegalArgumentException("${operator}는 사칙 연산기호가 아닙니다.")
            }
        }
```
이렇게 바꿀 수도 있어.

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,26 @@
+class StringCalculator {
+    fun calculate(keyboardInputString: String): Int {
+        val splitList = keyboardInputString.split(" ")
+        CalculateChecker().checkCalculable(splitList)
```
분리 하려는 노력이 보여서 뿌듯합니다 😀 아직 더 분리할 수 있다고 봐!!
"계산하는" 메서드에 문자열 분리, 검증로직, 계산이 다 들어가 있으니까!

여기서 검증하는 로직만 뺀거같은데, 검증은 어떤 시점에서 해야했던 걸까??
들어온 문자열이 유효한 수식인지 검증을 하는거잖아?
그렇다면, 잘 분리하기 위해 검증을 한건 아닐까?
더 생각해 본다면, 문자열을 분리하는 것과 그 수식이 맞는지 검증하는 것은 아예 다른 기능일까?

비슷한 기능끼리 똘똘 뭉쳐있고 협력을 긴밀하게 하고있다는 것을 응집도가 높다 라는 표현을 써.
지금은 응집도가 높은 편일까?

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -0,0 +1,26 @@
+        CalculateChecker().checkCalculable(splitList)
+        var resultNum = splitList[0].toInt()
+        for (i in 1 until splitList.size step 2) {
+            when (splitList[i]) {
+                "+" -> {
+                    resultNum += splitList[i + 1].toInt()
+                }
```
splitList[i] 이 뭔지, splitList[i + 1]이 뭔지 우리는 요구사항을 하도 많이 봐서 알거같은데.
처음 보는 사람이 이게 뭔지 알까?
val operator = splitList[i] 이런식으로 한 번 래핑하는편이 코드 한 줄을 늘리지만, 가독성을 높이는 방법일 수 있어.

SONPYEONGHWA replied:
감사합니다 조금 더 직관적인 코드로 바꿔보겠습니다 ㅎㅎ

malibinYun commented on src/test/kotlin/CalculatorCheckerTest.kt:
```diff
@@ -0,0 +1,15 @@
+
+class CalculatorCheckerTest {
+    @Test
+    fun checkCalculableTest() {
+        Assertions.assertThatThrownBy { CalculateChecker().checkCalculable(listOf("&")) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("사칙 연산기호가 아닙니다.")
+        Assertions.assertThatThrownBy { CalculateChecker().checkCalculable(listOf("1", "*", "/", "5")) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("연산자")
+    }
```
서로 다른 특징에 대해서 테스트를 하고있다고 생각해.
사칙 기호가 아닌놈이 들어왔을 때, 유효한 수식이 아닐 때.
각 다른 테스트로 분리하는게 설명서로 만들기에 더 적합하다고 생각해!

SONPYEONGHWA replied:
아하 분리해보겠습니다!!

malibinYun reviewed (changes requested):
피드백 반영 고생했습니다. 내 의도를 잘 알아준거 같아서 뿌듯합니다. 조금만 더 해봅시다잉

engineer-choi committed:
[2차 리팩터링 완료]
연산자 검사에 대한 불필요한 낭비 줄이기 위해 index 2개씩 검사하는 로직 적용.
calculate()의 내부에 있던 분리, 검증 과정 분리.
피드백 반영하여 전체적 리팩터링

jinsu4755 commented on src/main/kotlin/Main.kt:
```diff
@@ -0,0 +1,9 @@
+import java.lang.IllegalArgumentException
+
+fun main() {
+    val keyboardInputString = readLine() ?: throw IllegalArgumentException("아무것도 입력하지 않았습니다.")
```
그냥 개인적으로 궁금한 점인데 이 경우에 사용자가 아무것도 입력안한경우의 테스트가 존재하지 않아도 되는건가요??

malibinYun replied:
내가 알기로 readLine에서 null이 떨어지는 경우는 아무것도 입력하지 않았을 때가 아니라 다른 종류의 내부적인 무언가로 알고있어

chaeppy commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -1,22 +1,14 @@
-        val splitList = keyboardInputString.split(" ")
-        CalculateChecker().checkCalculable(splitList)
+    fun calculate(splitList: List<String>): Int {
         var resultNum = splitList[0].toInt()
```
CalculateChecker의 checkCalculable함수에서 spiltList의 첫 요소는 검사하지 않는 것 처럼 보이는데, 혹시 이부분에 숫자말고 다른 값을 넣어도 예외처리가 잘 되나요?

chaeppy commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -1,24 +1,31 @@
 import java.lang.IllegalArgumentException
 
 class CalculateChecker {
```
오오 이렇게 체크하는 클래스를 따로 만들 수도 있겠군여 ~~ !! 참고해야겠습니다 ㅎㅎ
근데 Calculate는 계산하다라는 뜻이라서 뭐를 검사하는 클래스인지..? 좀더 구체적으로 클래스명을 바꿔줘도 좋을 것 같다는 생각이 들기도 합니다..!! ((개인적인 의견입니다))

chaeppy commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -1,24 +1,31 @@
-    private fun checkOperatorNumber(operatorList: List<String>, numberList: List<String>) {
-        require(operatorList.size + 1 == numberList.size) { IllegalArgumentException("연산자 : $operatorList 숫자 : $numberList") }
+    fun checkSize(splitList: List<String>) {
+        require(splitList.size > 2) { IllegalArgumentException("계산 식이 아닙니다.\n 입력 받은 값 : $splitList") }
```
단순히 사이즈만 확인하는 함수 인 것 같은데, 어떤 Exception 인지 설명해주는 부분을 계산식이 아니라는 것 보다 조금 자세하게 수정해도 괜찮을 것 같다는 생각이 듭니다!!

chaeppy commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -1,24 +1,31 @@
+    }
+
+    fun checkNum(number: String) {
+        require(number.toIntOrNull() is Int) {
```
오 저는 try-catch문을 쓰려고했는데 이런 방법도 있군요 !! 좋은 것 같아요 !

malibinYun replied:
으음 내생각엔 try catch가 좀 더 나아보여!

chaeppy commented on src/test/kotlin/CalculatorCheckerTest.kt:
```diff
@@ -1,15 +1,40 @@
 class CalculatorCheckerTest {
+
     @Test
     fun checkCalculableTest() {
```
요거 checkSize, checkOperator 등 따로따로 검사해두 좋을 것 같아요 !!

chaeppy reviewed (commented):
수고하셨습니다!!! ^.^

malibinYun commented on src/test/kotlin/CalculatorCheckerTest.kt:
```diff
@@ -0,0 +1,40 @@
+
+class CalculatorCheckerTest {
+
+    @Test
+    fun checkCalculableTest() {
+        assertThatThrownBy { CalculateChecker().checkCalculable(listOf("1", "*", "9", "/", "3", "*", "*")) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("숫자가 입력되어야 합니다.")
+        assertThatThrownBy { CalculateChecker().checkCalculable(listOf("1", "*", "9", "/", "3", "&", "7")) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("사칙연산 기호만 입력할 수 있습니다.")
+        assertThatThrownBy { CalculateChecker().checkCalculable(listOf("1", "*")) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("계산 식이 아닙니다.")
+    }
```
요 테스트는 지워도 될거같아!

malibinYun commented on src/main/kotlin/CalculateChecker.kt:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    fun checkSize(splitList: List<String>) {
+        require(splitList.size > 2) { IllegalArgumentException("계산 식이 아닙니다.\n 입력 받은 값 : $splitList") }
+    }
+
+    fun checkNum(number: String) {
+        require(number.toIntOrNull() is Int) {
+            IllegalArgumentException("숫자가 입력되어야 합니다. \n 입력받은 값 : $number")
+        }
+    }
+
+    fun checkOperator(operator: String) {
+        require(FOUR_RULES_CALCULATION_LIST.contains(operator)) {
+            IllegalArgumentException("해당 자리에 사칙연산 기호만 입력할 수 있습니다. \n 입력받은 값 : $operator")
+        }
```
이녀석들은 외부에 공개할 필요가 없는 애들인거같아.

malibinYun reviewed (approved):
고생 많았습니다. 죠금 아쉽지만 미션 머지할게요~~
내가 생각했던건, StringSpliter 이런 객체에서 원시 String을 넘겨주면, 거기에서 검증을 모두 다 한 뒤, split된 List를 넘겨주는 그런 그림을 그리면서 피드백 해준거였어 !! 그래도 많이 발전한거 같아서 뿌듯합니다. 이건 안해두되구 참고해줘~~
이번 피드백들이랑 내일 가르쳐 줄 것들 같이 피드백 반영하면 될거같아. 고생했어~~

malibinYun commented:
아놔 당신
 CHJUNE-android  <- CHJUNE-android-step1
이렇게 풀리퀘 다시 날리세요! 바로 머지하겠습니다
